File: hot100/Solution0148.cpp
#include <queue>
#include <algorithm>
using namespace std;
// 给你链表的头结点 head ，请将其按 升序 排列并返回 排序后的链表 。

// 示例 1：

// 输入：head = [4,2,1,3]
// 输出：[1,2,3,4]
// 示例 2：

// 输入：head = [-1,5,3,4,0]
// 输出：[-1,0,3,4,5]
// 示例 3：

// 输入：head = []
// 输出：[]

// 提示：

// 链表中节点的数目在范围 [0, 5 * 104] 内
// -105 <= Node.val <= 105

// 进阶：你可以在 O(n log n) 时间复杂度和常数级空间复杂度下，对链表进行排序吗？
// 方法一：小根堆，优先级队列
class Solution
{
private:
    struct ListNode
    {
        int val;
        ListNode *next;
        ListNode(int x) : val(x), next(nullptr) {}
        ListNode(int x, ListNode *next) : val(x), next(next) {}
    };

public:
    ListNode *sortList(ListNode *head)
    {
        auto cmp = [](ListNode *a, ListNode *b)
        {
            return a->val > b->val;
        };
        priority_queue<ListNode *, vector<ListNode *>, decltype(cmp)> pq(cmp);
        for (ListNode *cur = head; cur != nullptr; cur = cur->next)
        {
            pq.push(cur);
        }
        ListNode *dummy = new ListNode(0);
        ListNode *tail = dummy;
        while (!pq.empty())
        {
            tail->next = pq.top();
            pq.pop();
            tail = tail->next;
        }
        tail->next = nullptr;
        return dummy->next;
    }
};

// 方法二：自顶向下的归并，空间复杂度是o(log n)
class Solution2
{
private:
    struct ListNode
    {
        int val;
        ListNode *next;
        ListNode(int x) : val(x), next(nullptr) {}
        ListNode(int x, ListNode *next) : val(x), next(next) {}
    };

public:
    ListNode *sortList(ListNode *head)
    {
        return sort(head, nullptr);
    }

private:
    // 处理[head,tail)区间
    ListNode *sort(ListNode *head, ListNode *tail)
    {
        if (!head)
            return nullptr;
        if (head->next == tail)
        {
            head->next = nullptr;
            return head;
        }
        ListNode *slow = head, *fast = head;
        while (fast != tail && fast->next != tail)
        {
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode *mid = slow;
        return merge(sort(head, mid), sort(mid, tail));
    }
    ListNode *merge(ListNode *head1, ListNode *head2)
    {
        ListNode *dummy_head = new ListNode(0);
        ListNode *dummy_tail = dummy_head;
        ListNode *pointer1 = head1, *pointer2 = head2;
        while (pointer1 != nullptr && pointer2 != nullptr)
        {
            if (pointer1->val > pointer2->val)
            {
                dummy_tail->next = pointer2;
                pointer2 = pointer2->next;
            }
            else
            {
                dummy_tail->next = pointer1;
                pointer1 = pointer1->next;
            }
            dummy_tail = dummy_tail->next;
        }
        dummy_tail->next = pointer1 != nullptr ? pointer1 : pointer2;
        return dummy_head->next;
    }
};

// 方法三：自底向上的归并，空间复杂度是O(1)
class Solution3
{
private:
    struct ListNode
    {
        int val;
        ListNode *next;
        ListNode(int x) : val(x), next(nullptr) {}
        ListNode(int x, ListNode *next) : val(x), next(next) {}
    };

public:
    ListNode *sortList(ListNode *head)
    {
        if (!head || !head->next)
        {
            return head;
        }
        int length = 0;
        for (ListNode *i = head; head != nullptr; i = i->next)
        {
            length++;
        }
        ListNode *dummy = new ListNode(0, head);
        for (int size = 1; size < length; size *= 2)
        {
            ListNode *prev = dummy;       // 已经排好序的链表的末尾
            ListNode *curr = dummy->next; // 需要排序的链表的开头
            while (curr != nullptr)
            {
                ListNode *left = curr;
                ListNode *right = split(curr, size);
                curr = split(right, size);
                ListNode *mergerd = merge(left, right); // 归并后的链表的链头
                prev->next = mergerd;
                // 一直把prev移动到归并后的链表的链尾
                while (prev->next != nullptr)
                {
                    prev = prev->next;
                }
            }
        }
    }

private:
    // 返回开头过size后，截断，给出第二个链表的开头
    ListNode *split(ListNode *head, int size)
    {
        if (!head)
            return nullptr;
        ListNode *current = head;
        for (int i = 1; i < size && current != nullptr; i++)
        {
            current = current->next;
        }
        // 说明当前链条的长度小于size
        if (current == nullptr)
        {
            return nullptr;
        }
        ListNode *next_list_head = current->next;
        current->next = nullptr; // 断开链接
        return next_list_head;
    }
    // 两个链表，归并
    ListNode *merge(ListNode *left, ListNode *right)
    {
        ListNode *dummy = new ListNode(0);
        ListNode *prev = dummy;
        ListNode *pointer1 = left, *pointer2 = right;
        while (pointer1 != nullptr && pointer2 != nullptr)
        {
            if (pointer1->val > pointer2->val)
            {
                prev->next = pointer2;
                pointer2 = pointer2->next;
            }
            else
            {
                prev->next = pointer1;
                pointer1 = pointer1->next;
            }
            prev = prev->next;
        }
        prev->next = (pointer1 != nullptr) ? pointer1 : pointer2;
        return dummy->next;
    }
};

File: hot100/Solution0148.java
package hot100;

// 给你链表的头结点 head ，请将其按 升序 排列并返回 排序后的链表 。

import java.util.PriorityQueue;

public class Solution0148 {
    private static class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode sortList(ListNode head) {
        PriorityQueue<ListNode> minHeap = new PriorityQueue<>((a, b) -> Integer.compare(a.val, b.val));
        ListNode current = head;
        while (current != null) {
            minHeap.offer(current);
            current = current.next;
        }
        ListNode dummy = new ListNode(0);
        current = dummy;
        while (!minHeap.isEmpty()) {
            current.next = minHeap.poll();
            current = current.next;
        }
        current.next = null; // 断开最后一个节点的 next 指针，避免形成循环
        return dummy.next;
    }

    public ListNode sortList2(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode slow = head, fast = head.next;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode mid = slow.next;
        slow.next = null; // 断开链表，将链表分为两部分
        ListNode left = sortList2(head);
        ListNode right = sortList2(mid);
        return merge(left, right);
    }

    private ListNode merge(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(0);
        ListNode current = dummy;
        while (l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                current.next = l1;
                l1 = l1.next;
            } else {
                current.next = l2;
                l2 = l2.next;
            }
            current = current.next;
        }
        current.next = (l1 != null) ? l1 : l2;
        return dummy.next;
    }

    public ListNode sortList3(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        int length = 0;
        ListNode current = head;
        while (current != null) {
            length++;
            current = current.next;
        }
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        for (int step = 1; step < length; step *= 2) {
            ListNode prev = dummy, curr = dummy.next;
            while (curr != null) {
                ListNode left = curr;
                ListNode right = split(left, step);
                curr = split(right, step);
                prev.next = merge(left, right);
                while (prev.next != null) {
                    prev = prev.next;
                }
            }
        }
        return dummy.next;
    }

    private ListNode split(ListNode head, int step) {
        for (int i = 1; head != null && i < step; i++) {
            head = head.next;
        }
        if (head == null) {
            return null;
        }
        ListNode second = head.next;
        head.next = null; // 断开链表
        return second;
    }

    public static void main(String[] args) {
        Solution0148 solution = new Solution0148();
        ListNode head = new ListNode(4, new ListNode(2, new ListNode(1, new ListNode(3))));
        ListNode sortedHead = solution.sortList3(head);
        while (sortedHead != null) {
            System.out.print(sortedHead.val + " ");
            sortedHead = sortedHead.next;
        }
    }
}
